src/game/objects/creatures/ape.c

```diff
@@ -48,77 +48,83 @@ typedef enum {
-        return;
-    }
-    obj->initialise = Creature_Initialise;
-    obj->control = Ape_Control;
-    obj->collision = Creature_Collision;
-    obj->shadow_size = UNIT_SHADOW / 2;
-    obj->hit_points = APE_HITPOINTS;
-    obj->pivot_length = 250;
-    obj->radius = APE_RADIUS;
-    obj->smartness = APE_SMARTNESS;
-    obj->intelligent = 1;
-    obj->save_position = 1;
-    obj->save_hitpoints = 1;
-    obj->save_anim = 1;
-    obj->save_flags = 1;
-    g_AnimBones[obj->bone_index + 52] |= BEB_ROT_Y;
-}
+static bool Ape_Vault(int16_t item_num, int16_t angle);
 
-void Ape_Vault(int16_t item_num, int16_t angle)
+static bool Ape_Vault(int16_t item_num, int16_t angle)
 {
     ITEM_INFO *item = &g_Items[item_num];
     CREATURE_INFO *ape = item->data;
+    int32_t x = item->pos.x >> WALL_SHIFT;
+    int32_t y = item->pos.y;
+    int32_t z = item->pos.z >> WALL_SHIFT;
+    int16_t room_number = item->room_number;
 
     if (ape->flags & APE_TURN_L_FLAG) {
         item->pos.y_rot -= PHD_90;
         ape->flags &= ~APE_TURN_L_FLAG;
-    } else if (item->flags & APE_TURN_R_FLAG) {
+    } else if (ape->flags & APE_TURN_R_FLAG) {
         item->pos.y_rot += PHD_90;
         ape->flags &= ~APE_TURN_R_FLAG;
     }
 
-    int32_t xx = item->pos.z >> WALL_SHIFT;
-    int32_t yy = item->pos.x >> WALL_SHIFT;
-    int32_t y = item->pos.y;
-
     Creature_Animate(item_num, angle, 0);
 
     if (item->pos.y > y - STEP_L * 3 / 2) {
-        return;
+        return false;
     }
 
-    int32_t x_floor = item->pos.z >> WALL_SHIFT;
-    int32_t y_floor = item->pos.x >> WALL_SHIFT;
-    if (xx == x_floor) {
-        if (yy == y_floor) {
-            return;
+    int32_t x_floor = item->pos.x >> WALL_SHIFT;
+    int32_t z_floor = item->pos.z >> WALL_SHIFT;
+
+    if (z == z_floor) {
+        if (x == x_floor) {
+            return false;
         }
 
-        if (yy < y_floor) {
-            item->pos.x = (y_floor << WALL_SHIFT) - APE_SHIFT;
-            item->pos.y_rot = PHD_90;
-        } else {
-            item->pos.x = (yy << WALL_SHIFT) + APE_SHIFT;
+        if (x >= x_floor) {
             item->pos.y_rot = -PHD_90;
+            item->pos.x = (x << WALL_SHIFT) + APE_SHIFT;
+        } else {
+            item->pos.y_rot = PHD_90;
+            item->pos.x = (x_floor << WALL_SHIFT) - APE_SHIFT;
         }
-    } else if (yy == y_floor) {
-        if (xx < x_floor) {
-            item->pos.z = (x_floor << WALL_SHIFT) - APE_SHIFT;
+    } else if (x == x_floor) {
+        if (z < z_floor) {
             item->pos.y_rot = 0;
+            item->pos.z = (z_floor << WALL_SHIFT) - APE_SHIFT;
         } else {
-            item->pos.z = (xx << WALL_SHIFT) + APE_SHIFT;
             item->pos.y_rot = -PHD_180;
+            item->pos.z = (z << WALL_SHIFT) + APE_SHIFT;
         }
     }
 
+    item->floor = y;
     item->pos.y = y;
-    item->current_anim_state = APE_VAULT;
-    Item_SwitchToAnim(item, APE_VAULT_ANIM, -1);
+
+    Item_NewRoom(item_num, room_number);
```

Comment:
Shouldn't we have `if (item->room_number != room_number)`?
Can we rename `room_number` to `room_num`?

Reply:
I can rename. Um, maybe? I just left it like this to make it consistent with later TRs. The vault function in TR3+ checks `if (vault)` before `Item_NewRoom`. However, the `Ape_Vault` is slightly different since we don't want have a downward vault animation. So, the function only gets to this part at the end and returns true if the ape is performing a vault up.

Reply:
I think so too, but I'd like to stay consistent with the other places that use Item_NewRoom.